File: framec_tests/golang/match/match.go
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files


func NewMatch() Match {
    m := &matchStruct{}
    
    // Validate interfaces
    var _ Match = m
    var _ Match_actions = m
    
    // Create and intialize start state compartment.
    m._compartment_ = NewMatchCompartment(MatchState_Init)
    
    // Override domain variables.
    m.tape = []string{}
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type MatchState uint

const (
    MatchState_Init MatchState = iota
    MatchState_EmptyMatch
    MatchState_SimpleMatch
    MatchState_MultiMatch
    MatchState_NestedMatch
    MatchState_ChildMatch
    MatchState_Final
)

type Match interface {
    Empty() 
    Simple() 
    Multi() 
    Nested() 
    Child() 
    OnInt(i int) 
    Onstring(s string) 
}

type Match_actions interface {
    log(msg string) 
}


type matchStruct struct {
    _compartment_ *MatchCompartment
    _nextCompartment_ *MatchCompartment
    tape []string
}

//===================== Interface Block ===================//

func (m *matchStruct) Empty()  {
    e := framelang.FrameEvent{Msg:"Empty"}
    m._mux_(&e)
}

func (m *matchStruct) Simple()  {
    e := framelang.FrameEvent{Msg:"Simple"}
    m._mux_(&e)
}

func (m *matchStruct) Multi()  {
    e := framelang.FrameEvent{Msg:"Multi"}
    m._mux_(&e)
}

func (m *matchStruct) Nested()  {
    e := framelang.FrameEvent{Msg:"Nested"}
    m._mux_(&e)
}

func (m *matchStruct) Child()  {
    e := framelang.FrameEvent{Msg:"Child"}
    m._mux_(&e)
}

func (m *matchStruct) OnInt(i int)  {
    params := make(map[string]interface{})
    params["i"] = i
    e := framelang.FrameEvent{Msg:"OnInt", Params:params}
    m._mux_(&e)
}

func (m *matchStruct) Onstring(s string)  {
    params := make(map[string]interface{})
    params["s"] = s
    e := framelang.FrameEvent{Msg:"Onstring", Params:params}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *matchStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case MatchState_Init:
        m._MatchState_Init_(e)
    case MatchState_EmptyMatch:
        m._MatchState_EmptyMatch_(e)
    case MatchState_SimpleMatch:
        m._MatchState_SimpleMatch_(e)
    case MatchState_MultiMatch:
        m._MatchState_MultiMatch_(e)
    case MatchState_NestedMatch:
        m._MatchState_NestedMatch_(e)
    case MatchState_ChildMatch:
        m._MatchState_ChildMatch_(e)
    case MatchState_Final:
        m._MatchState_Final_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *matchStruct) _MatchState_Init_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "Empty":
        compartment := NewMatchCompartment(MatchState_EmptyMatch)
        m._transition_(compartment)
        
        return
    case "Simple":
        compartment := NewMatchCompartment(MatchState_SimpleMatch)
        m._transition_(compartment)
        
        return
    case "Multi":
        compartment := NewMatchCompartment(MatchState_MultiMatch)
        m._transition_(compartment)
        
        return
    case "Nested":
        compartment := NewMatchCompartment(MatchState_NestedMatch)
        m._transition_(compartment)
        
        return
    case "Child":
        compartment := NewMatchCompartment(MatchState_ChildMatch)
        m._transition_(compartment)
        
        return
    }
}

func (m *matchStruct) _MatchState_EmptyMatch_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "Onstring":
        if e.Params["s"].(string) == "" || e.Params["s"].(string) == "foo" {
            m.log("empty")
        } else {
            m.log("?")
        }
        
        return
    }
}  //  TODO: matching only the empty string is broken


func (m *matchStruct) _MatchState_SimpleMatch_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnInt":
        if e.Params["i"].(int) == 0 {
            m.log("0")
        } else if e.Params["i"].(int) == 42 {
            m.log("42")
        } else if e.Params["i"].(int) == 42 {
            m.log("!!!")
        } else if e.Params["i"].(int) == -200 {
            m.log("-200")
        } else {
            m.log("?")
        }
        
        return
    case "Onstring":
        if e.Params["s"].(string) == "hello" {
            m.log("hello")
        } else if e.Params["s"].(string) == "hello" {
            m.log("!!!")
        } else if e.Params["s"].(string) == "goodbye" {
            m.log("goodbye")
        } else if e.Params["s"].(string) == "Testing 1, 2, 3..." {
            m.log("testing")
        } else if e.Params["s"].(string) == "$10!" {
            m.log("money")
        } else {
            m.log("?")
        }
        
        return
    }
}

func (m *matchStruct) _MatchState_MultiMatch_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnInt":
        if e.Params["i"].(int) == 3 || e.Params["i"].(int) == -7 {
            m.log("3|-7")
        } else if e.Params["i"].(int) == -4 || e.Params["i"].(int) == 5 || e.Params["i"].(int) == 6 {
            m.log("-4|5|6")
        } else {
            m.log("?")
        }
        
        return
    case "Onstring":
        if e.Params["s"].(string) == "$10" || e.Params["s"].(string) == "12.5%" || e.Params["s"].(string) == "@#*!" {
            m.log("symbols")
        } else if e.Params["s"].(string) == " " || e.Params["s"].(string) == "  " || e.Params["s"].(string) == "\t" || e.Params["s"].(string) == "\n" {
            m.log("whitespace")
        } else {
            m.log("?")
        }
        
        return
    }
}

func (m *matchStruct) _MatchState_NestedMatch_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnInt":
        if e.Params["i"].(int) > 0 {
            if e.Params["i"].(int) == 1 || e.Params["i"].(int) == 2 || e.Params["i"].(int) == 3 {
                m.log("1-3")
                if e.Params["i"].(int) == 1 {
                    m.log("1")
                } else if e.Params["i"].(int) == 2 {
                    m.log("2")
                } else {
                    m.log("3")
                }
            } else if e.Params["i"].(int) == 4 || e.Params["i"].(int) == 5 {
                m.log("4-5")
                if e.Params["i"].(int) == 4 {
                    m.log("4")
                } else {
                    m.log("5")
                }
            } else {
                m.log("too big")
            }
        } else {
            m.log("too small")
        }
        
        return
    case "Onstring":
        if e.Params["s"].(string) == "hello" || e.Params["s"].(string) == "hola" || e.Params["s"].(string) == "bonjour" {
            m.log("greeting")
            if e.Params["s"].(string) == "hello" {
                m.log("English")
            } else if e.Params["s"].(string) == "hola" {
                m.log("Spanish")
            } else {
                m.log("French")
            }
        } else if e.Params["s"].(string) == "goodbye" || e.Params["s"].(string) == "adios" || e.Params["s"].(string) == "au revoir" {
            m.log("farewell")
            if e.Params["s"].(string) == "goodbye" {
                m.log("English")
            } else if e.Params["s"].(string) == "adios" {
                m.log("Spanish")
            } else {
                m.log("French")
            }
        } else {
            m.log("?")
        }
        
        return
    }
}

func (m *matchStruct) _MatchState_ChildMatch_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnInt":
        if e.Params["i"].(int) == 0 {
            compartment := NewMatchCompartment(MatchState_Final)
            m._transition_(compartment)
            return
        } else if e.Params["i"].(int) == 3 {
            m.log("3")
        } else if e.Params["i"].(int) == 4 {
            m.log("4")
            
            return
        } else if e.Params["i"].(int) == 42 {
            m.log("42 in child")
        } else if e.Params["i"].(int) == 5 {
            m.log("5")
            compartment := NewMatchCompartment(MatchState_Final)
            m._transition_(compartment)
            return
        } else {
            m.log("no match in child")
        }
        
    case "Onstring":
        if e.Params["s"].(string) == "hello" {
            m.log("hello in child")
        } else if e.Params["s"].(string) == "goodbye" {
            compartment := NewMatchCompartment(MatchState_Final)
            m._transition_(compartment)
            return
        } else if e.Params["s"].(string) == "Testing 1, 2, 3..." {
            m.log("testing in child")
            
            return
        } else {
            m.log("no match in child")
        }
        
    }
    m._MatchState_SimpleMatch_(e)
    
}

func (m *matchStruct) _MatchState_Final_(e *framelang.FrameEvent) {
    switch e.Msg {
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *matchStruct) _transition_(compartment *MatchCompartment) {
    m._nextCompartment_ = compartment
}

func (m *matchStruct) _do_transition_(nextCompartment *MatchCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions

func (m *matchStruct) log(msg string)  {}

********************************************************/

//=============== Compartment ==============//

type MatchCompartment struct {
    State MatchState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewMatchCompartment(state MatchState) *MatchCompartment {
    c := &MatchCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}

File: framec_tests/python/src/match/match.py
#Emitted from framec_v0.11.0



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class Match:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and intialize start state compartment.
        
        self.__compartment: 'MatchCompartment' = MatchCompartment('__match_state_Init')
        self.__next_compartment: 'MatchCompartment' = None
        self.__compartment: MatchCompartment = MatchCompartment(self.__state)
        self.__next_compartment: MatchCompartment = None
        
        # Initialize domain
        
        self.tape  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def Empty(self,):
        __e = FrameEvent("Empty",None)
        self.__kernel(__e)
    
    def Simple(self,):
        __e = FrameEvent("Simple",None)
        self.__kernel(__e)
    
    def Multi(self,):
        __e = FrameEvent("Multi",None)
        self.__kernel(__e)
    
    def Nested(self,):
        __e = FrameEvent("Nested",None)
        self.__kernel(__e)
    
    def Child(self,):
        __e = FrameEvent("Child",None)
        self.__kernel(__e)
    
    def OnInt(self,i: int):
        parameters = {}
        parameters["i"] = i
        __e = FrameEvent("OnInt",parameters)
        self.__kernel(__e)
    
    def Onstring(self,s: str):
        parameters = {}
        parameters["s"] = s
        __e = FrameEvent("Onstring",parameters)
        self.__kernel(__e)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $Init
    
    def __match_state_Init(self, __e):
        if __e._message == "Empty":
            next_compartment = MatchCompartment('__match_state_EmptyMatch')
            self.__transition(next_compartment)
            return
        elif __e._message == "Simple":
            next_compartment = MatchCompartment('__match_state_SimpleMatch')
            self.__transition(next_compartment)
            return
        elif __e._message == "Multi":
            next_compartment = MatchCompartment('__match_state_MultiMatch')
            self.__transition(next_compartment)
            return
        elif __e._message == "Nested":
            next_compartment = MatchCompartment('__match_state_NestedMatch')
            self.__transition(next_compartment)
            return
        elif __e._message == "Child":
            next_compartment = MatchCompartment('__match_state_ChildMatch')
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $EmptyMatch
    
    def __match_state_EmptyMatch(self, __e):
        if __e._message == "Onstring":
            if ((__e._parameters["s"] == "") or (__e._parameters["s"] == "foo")):
                self.log_do("empty")
            
            else:
                self.log_do("?")
            
            return
      #  TODO: matching only the empty string is broken
    
    
    # ----------------------------------------
    # $SimpleMatch
    
    def __match_state_SimpleMatch(self, __e):
        if __e._message == "OnInt":
            if (__e._parameters["i"] == 0):
                self.log_do("0")
            elif (__e._parameters["i"] == 42):
                self.log_do("42")
            elif (__e._parameters["i"] == 42):
                self.log_do("!!!")
            elif (__e._parameters["i"] == -200):
                self.log_do("-200")
            
            else:
                self.log_do("?")
            
            return
        elif __e._message == "Onstring":
            if ((__e._parameters["s"] == "hello")):
                self.log_do("hello")
            elif ((__e._parameters["s"] == "hello")):
                self.log_do("!!!")
            elif ((__e._parameters["s"] == "goodbye")):
                self.log_do("goodbye")
            elif ((__e._parameters["s"] == "Testing 1, 2, 3...")):
                self.log_do("testing")
            elif ((__e._parameters["s"] == "$10!")):
                self.log_do("money")
            
            else:
                self.log_do("?")
            
            return
    
    # ----------------------------------------
    # $MultiMatch
    
    def __match_state_MultiMatch(self, __e):
        if __e._message == "OnInt":
            if (__e._parameters["i"] == 3) or (__e._parameters["i"] == -7):
                self.log_do("3|-7")
            elif (__e._parameters["i"] == -4) or (__e._parameters["i"] == 5) or (__e._parameters["i"] == 6):
                self.log_do("-4|5|6")
            
            else:
                self.log_do("?")
            
            return
        elif __e._message == "Onstring":
            if ((__e._parameters["s"] == "$10") or (__e._parameters["s"] == "12.5%") or (__e._parameters["s"] == "@#*!")):
                self.log_do("symbols")
            elif ((__e._parameters["s"] == " ") or (__e._parameters["s"] == "  ") or (__e._parameters["s"] == "\t") or (__e._parameters["s"] == "\n")):
                self.log_do("whitespace")
            
            else:
                self.log_do("?")
            
            return
    
    # ----------------------------------------
    # $NestedMatch
    
    def __match_state_NestedMatch(self, __e):
        if __e._message == "OnInt":
            if  __e._parameters["i"] > 0:
                if (__e._parameters["i"] == 1) or (__e._parameters["i"] == 2) or (__e._parameters["i"] == 3):
                    self.log_do("1-3")
                    if (__e._parameters["i"] == 1):
                        self.log_do("1")
                    elif (__e._parameters["i"] == 2):
                        self.log_do("2")
                    
                    else:
                        self.log_do("3")
                    
                elif (__e._parameters["i"] == 4) or (__e._parameters["i"] == 5):
                    self.log_do("4-5")
                    if  __e._parameters["i"] == 4:
                        self.log_do("4")
                    else:
                        self.log_do("5")
                    
                
                else:
                    self.log_do("too big")
                
            else:
                self.log_do("too small")
            
            return
        elif __e._message == "Onstring":
            if ((__e._parameters["s"] == "hello") or (__e._parameters["s"] == "hola") or (__e._parameters["s"] == "bonjour")):
                self.log_do("greeting")
                if ((__e._parameters["s"] == "hello")):
                    self.log_do("English")
                elif ((__e._parameters["s"] == "hola")):
                    self.log_do("Spanish")
                
                else:
                    self.log_do("French")
                
            elif ((__e._parameters["s"] == "goodbye") or (__e._parameters["s"] == "adios") or (__e._parameters["s"] == "au revoir")):
                self.log_do("farewell")
                if ((__e._parameters["s"] == "goodbye")):
                    self.log_do("English")
                elif ((__e._parameters["s"] == "adios")):
                    self.log_do("Spanish")
                
                else:
                    self.log_do("French")
                
            
            else:
                self.log_do("?")
            
            return
    
    # ----------------------------------------
    # $ChildMatch
    
    def __match_state_ChildMatch(self, __e):
        if __e._message == "OnInt":
            if (__e._parameters["i"] == 0):
                next_compartment = MatchCompartment('__match_state_Final')
                self.__transition(next_compartment)
                return
            elif (__e._parameters["i"] == 3):
                self.log_do("3")
            elif (__e._parameters["i"] == 4):
                self.log_do("4")
                
                return
            elif (__e._parameters["i"] == 42):
                self.log_do("42 in child")
            elif (__e._parameters["i"] == 5):
                self.log_do("5")
                next_compartment = MatchCompartment('__match_state_Final')
                self.__transition(next_compartment)
                return
            
            else:
                self.log_do("no match in child")
            
        elif __e._message == "Onstring":
            if ((__e._parameters["s"] == "hello")):
                self.log_do("hello in child")
            elif ((__e._parameters["s"] == "goodbye")):
                next_compartment = MatchCompartment('__match_state_Final')
                self.__transition(next_compartment)
                return
            elif ((__e._parameters["s"] == "Testing 1, 2, 3...")):
                self.log_do("testing in child")
                
                return
            
            else:
                self.log_do("no match in child")
            
        
        self.__match_state_SimpleMatch(__e)
        
    
    # ----------------------------------------
    # $Final
    
    def __match_state_Final(self, __e):
        pass
        
    
    # ===================== Actions Block =================== #
    
    def log_do(self,msg: str):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__match_state_Init':
            self.__match_state_Init(__e)
        elif self.__compartment.state == '__match_state_EmptyMatch':
            self.__match_state_EmptyMatch(__e)
        elif self.__compartment.state == '__match_state_SimpleMatch':
            self.__match_state_SimpleMatch(__e)
        elif self.__compartment.state == '__match_state_MultiMatch':
            self.__match_state_MultiMatch(__e)
        elif self.__compartment.state == '__match_state_NestedMatch':
            self.__match_state_NestedMatch(__e)
        elif self.__compartment.state == '__match_state_ChildMatch':
            self.__match_state_ChildMatch(__e)
        elif self.__compartment.state == '__match_state_Final':
            self.__match_state_Final(__e)
        
    def __transition(self, next_compartment: 'MatchCompartment'):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class MatchCompartment:

    def __init__(self,state):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
    